component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -155,34 +164,54 @@ protected void initiateAuthenticationRequest(HttpServletRequest request,
             IdentityHelperUtil
                     .loadApplicationAuthenticationXMLFromRegistry(context, getName(), tenantDomain);
         }
-        String retryParam = "";
+
+        AuthenticatedUser authenticatedUserFromContext = TOTPUtil.getAuthenticatedUser(context);
+        if (authenticatedUserFromContext == null) {
+            throw new AuthenticationFailedException(
+                    ErrorMessages.ERROR_CODE_NO_AUTHENTICATED_USER.getCode(),
+                    ErrorMessages.ERROR_CODE_NO_AUTHENTICATED_USER.getMessage());
+        }
+
+        /*
+        The username that the server is using to identify the user, is needed to be identified, as
+        for the federated users, the username in the authentication context may not be same as the
+        username when the user is provisioned to the server.
+         */
+        String mappedLocalUsername = getMappedLocalUsername(authenticatedUserFromContext, context);
+
+        /*
+        If the mappedLocalUsername is blank, that means this is an initial login attempt by an non provisioned
+        federated user.
+         */
+        boolean isInitialFederationAttempt = StringUtils.isBlank(mappedLocalUsername);
+
         try {
-            FederatedAuthenticatorUtil.setUsernameFromFirstStep(context);
-            username = String.valueOf(context.getProperty("username"));
-            authenticatedUser = (AuthenticatedUser) context.getProperty("authenticatedUser");
-            // find the authenticated user.
-            if (authenticatedUser == null) {
-                diagnosticLog.error("Cannot find an authenticated user in the context. Cannot proceed further " +
-                        "without identifying the user");
-                throw new AuthenticationFailedException(
-                        "Authentication failed!. Cannot proceed further without identifying the user");
-            }
+            AuthenticatedUser authenticatingUser = resolveAuthenticatingUser(context, authenticatedUserFromContext,
+                    mappedLocalUsername, tenantDomain, isInitialFederationAttempt);
+            username = UserCoreUtil.addTenantDomainToEntry(authenticatingUser.getUserName(), tenantDomain);
+            context.setProperty(TOTPAuthenticatorConstants.AUTHENTICATED_USER, authenticatingUser);
+
+            String retryParam = "";
             if (context.isRetrying()) {
                 retryParam = "&authFailure=true&authFailureMsg=login.fail.message";
             }
-            boolean isTOTPEnabled = isTOTPEnabledForLocalUser(username);
-            if (log.isDebugEnabled()) {
-                log.debug("TOTP is enabled by user: " + isTOTPEnabled);
+            boolean isTOTPEnabled =false;
+            // Not required to check the TOTP enable state for the initial login of the federated users.
+            if (!isInitialFederationAttempt) {
+                isTOTPEnabled = isTOTPEnabledForLocalUser(UserCoreUtil.addDomainToName(username,
```

Comment:
shall we change the variable name. it's confusing.

Reply:
Fixed https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/105/commits/9ea77f5434ce5d211d19c714fc2f09fadf192a88